### ISSUE TITLE
RUF045 added in Ruff 0.9.7 should ignore `ClassVar` fields

### ISSUE THREAD

**Issue by kkom** [astral-sh/ruff#16297]
### Description

According to Python's docs class variables should be ignored by the dataclass mechanism: https://docs.python.org/3/library/dataclasses.html#class-variables

Which means that I think they don't need an explicit type hint in cases like this:

```python
@dataclass
class MyBaseClass:
    class_var_field: ClassVar[int]

@dataclass
class MySubClass(MyBaseClass):
    class_var_field: 7
```

But ruff now catches `class_var_field: 7` as a lint violation.

**Comment by MichaReiser** [comment]
The part I don't see his how inheritance is relevant here. I can remove the base class and get the exact same fields for `MySubClass`:

```
... @dataclass
... class MySubClass:
...      class_var_field = 7
...      other_field: int = 8

__init__(self, other_field: int = 8) -> None
    Initialize self.  See help(type(self)) for accurate signature.
```

**Comment by MichaReiser** [comment]
I talked to @AlexWaygood because I couldn't find any reference for how type checkers handle `ClassVar` for "inherited" class attributes and he confirmed what you both point out, they handle such attributes as class vars. 

That means type checkers can clearly determine whether a field is an instance attribute or a class attribute. 


However, as my *not knowing* has shown, using explicit `ClassVar` annotations can help readers understand whether a field is an instance or class attribute without having to click through the entire class hierarchy. 

Reading through the original issue (https://github.com/astral-sh/ruff/issues/12877), I get the impression that the focus is mainly the second -- improve clarity for readers. In fact, the `ClassVar` attribute technically is never necessary. Attributes without an annotation always are class variables and using `ClassVar` is mainly to improve readability. That's why I think that making this exception for inherited attributes goes against the rule's intent.

@adigitoleo what's your take on this as the person who originally requested the addition of the rule?

**Comment by adigitoleo** [comment]
That's essentially correct. The rule was added in response to the issues I raised around the arguably surprising behaviour of un-annotated variables in dataclass definitions. There is a detailed pitch in the linked issue, and some discussion in the forum (link in the OP of that issue).

Although it is valid syntax, I found it to be a stumbling block, especially when trying to offer a dataclass as public API for (novice) users to subclass. Using the explicit annotation for class variables makes the behaviour obvious.

Put another way: `ClassVar` being optional implicitly makes type annotations for instance variables mandatory. There is no way to avoid using at least _some_ annotations with dataclass variables. It is therefore my opinion (shared, I hope, by those who merged the new rule) that the least surprising dataclass definition, for readers (& maintainers) will necessarily use annotations for all variables.

Note that there are also other rules in the RUF category that trigger on valid, but "distasteful" syntax.

**Comment by adigitoleo** [comment]
Oh, I see now that the point being discussed here is maybe more specifically this one:

> `ClassVar` annotations are inheritable in the type system, so an inherited one should be deemed equally explicit

I'm still leaning towards ruff taking the conservative stance that inherited `ClassVar` are not "equally explicit", because they do not explicitly appear in the subclass definition. However, I am sympathetic to the concern that this will add more verbosity in some cases ("duplication" of the `ClassVar` annotation on the subclass). My feeling is that the benefit of encouraging `ClassVar` annotations regardless of inheritance outweighs this concern. When using lots of inherited dataclass class variables in a codebase, I would prefer to switch off RUF045 in the linter configuration.

**Comment by kkom** [comment]
@adigitoleo – I see! If it is indeed true there is no functional difference between an "inherited" `ClassVar` and an "implied" one, I'm happy with biasing toward explicitness.
